src/Text/Parsing/StringParser/String.purs

```diff
@@ -111,3 +115,39 @@ anyLetter = lowerCaseChar <|> upperCaseChar <?> "Expected a letter"
 -- | Match a letter or a number.
 alphaNum :: Parser Char
 alphaNum = anyLetter <|> anyDigit <?> "Expected a letter or a number"
+
+-- | Build the regular expression from the pattern and match it, ensuring
+-- | that the pattern only attempts to match from the start of the target.
+regex' :: String -> Parser String
+regex' pat =
+    case Regex.regex pattern noFlags of
+      Left _ ->
+        fail $ "Text.Parsing.StringParser.String.regex': illegal regex " <> pat
+      Right r ->
+        regex r
+    where
+      pattern =
+        case stripPrefix (Pattern "^") pat of
+          Nothing ->
+            "^" <> pat
+          _ ->
+            pat
+
+-- | Match the regular expression.
+regex :: Regex.Regex -> Parser String
+regex r =
+  Parser \{ str, pos } ->
+    let
+      remainder = drop pos str
+    in
+      -- reduce the possible array of matches to 0 or 1 elements to aid Array pattern matching
+      case uncons $ fromMaybe [] $ Regex.match r remainder of
+        Just { head: Just matched, tail: _ }  ->
+          -- only accept matches at position 0
+          case stripPrefix (Pattern matched) remainder of
```

Comment:
Just because the string matches at the start doesn't mean it was the same match, right? Consider something like `a+(?!b)` applied to the string `aaabaaa`. There is a match, `aaa`, and the string starts with that string, but the parser should actually fail.

Maybe we should only export the `regex'` version you originally had then, unless we can find a way to get position information for each match.